cpp/test/sg/rf_test.cu

```diff
@@ -40,305 +106,282 @@ struct RfInputs {
-
-    // Populate data (assume Col major)
-    std::vector<T> data_h = {30.0, 1.0, 2.0, 0.0, 10.0, 20.0, 10.0, 40.0};
-    data_h.resize(data_len);
-    raft::update_device(data, data_h.data(), data_len, stream);
-
-    // Populate labels
-    labels_h = {0, 1, 0, 4};
-    labels_h.resize(params.n_rows);
-    preprocess_labels(params.n_rows, labels_h, labels_map);
-    raft::update_device(labels, labels_h.data(), params.n_rows, stream);
-
-    forest = new typename ML::RandomForestMetaData<T, int>;
-    null_trees_ptr(forest);
-
-    raft::handle_t handle(rf_params.n_streams);
-    handle.set_stream(stream);
-
-    fit(handle, forest, data, params.n_rows, params.n_cols, labels,
-        labels_map.size(), rf_params);
-
-    CUDA_CHECK(cudaStreamSynchronize(stream));
-    //print_rf_detailed(forest);
-    // Inference data: same as train, but row major
-    int inference_data_len = params.n_inference_rows * params.n_cols;
-    inference_data_h = {30.0, 10.0, 1.0, 20.0, 2.0, 10.0, 0.0, 40.0};
-    inference_data_h.resize(inference_data_len);
-    raft::allocate(inference_data_d, inference_data_len);
-    raft::update_device(inference_data_d, inference_data_h.data(),
-                        inference_data_len, stream);
-
-    predict(handle, forest, inference_data_d, params.n_inference_rows,
-            params.n_cols, predicted_labels);
-    // Predict and compare against known labels
-    RF_metrics tmp =
-      score(handle, forest, labels, params.n_inference_rows, predicted_labels);
-
-    CUDA_CHECK(cudaStreamSynchronize(stream));
-    CUDA_CHECK(cudaStreamDestroy(stream));
-    accuracy = tmp.accuracy;
+// Classification
+template <typename DataT, typename LabelT, typename RfT>
+void TrainScore(const raft::handle_t& handle, RfTestParams params, DataT* X,
+                DataT* X_transpose, LabelT* y, RfT* forest,
+                RF_metrics& metrics) {
+  RF_params rf_params = set_rf_params(
+    params.max_depth, params.max_leaves, params.max_features, params.n_bins,
+    params.min_samples_leaf, params.min_samples_split,
+    params.min_impurity_decrease, params.bootstrap, params.n_trees,
+    params.max_samples, 0, params.split_criterion, params.n_streams, 128);
+  if constexpr (std::is_integral_v<LabelT>) {
+    fit(handle, forest, X, params.n_rows, params.n_cols, y, params.n_labels,
+        rf_params);
+  } else {
+    fit(handle, forest, X, params.n_rows, params.n_cols, y, rf_params);
   }
 
-  void SetUp() override { basicTest(); }
-
-  void TearDown() override {
-    accuracy = -1.0f;  // reset accuracy
-    postprocess_labels(params.n_rows, labels_h, labels_map);
-    inference_data_h.clear();
-    labels_h.clear();
-    labels_map.clear();
+  thrust::device_vector<LabelT> pred(params.n_rows);
+  predict(handle, forest, X_transpose, params.n_rows, params.n_cols,
+          pred.data().get());
+  // Predict and compare against known labels
+  metrics = score(handle, forest, y, params.n_rows, pred.data().get());
```

Comment:
we seem `fit()` model on `X` and `y` and then we `predict()` for input `X_transpose` and score the `pred` with respect to `y`. 
Just a few doubts:
1. why `X_transpose`? 
2. if so, why is the resulted predictions scored with respect to `y` ? 
3. we expect the score to be bad right? how do we know for sure that a model with depth `d` is less bad than one with depth `d-1`?

I hope I'm not missing anything obvious 🤔

Reply:
https://github.com/rapidsai/cuml/blob/964f40ee426c43b5f7d9e48cca9aed9f05c805b6/cpp/src/randomforest/randomforest.cuh#L251

The fit function accepts col major data, and predict accepts row major.

Reply:
ah! thanks Rory, apologies for some unnecessary overthinking!